src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1336 @@
+    out[valid_mask] = array[valid_mask] <= threshold
+    return out
+
+def ratio_op(lulc_array, soil_group_array, ratio_lookup, sorted_lucodes):
+    """Make an array of stormwater retention or infiltration ratios from 
+    arrays of LULC codes and hydrologic soil groups.
+
+    Args:
+        lulc_array (numpy.ndarray): 2D array of LULC codes
+        soil_group_array (numpy.ndarray): 2D array with the same shape as
+            ``lulc_array``. Values in {1, 2, 3, 4} corresponding to soil 
+            groups A, B, C, and D.
+        ratio_lookup (numpy.ndarray): 2D array where rows correspond to 
+            sorted LULC codes and columns correspond to soil groups
+            A, B, C, D in order. Shape: (number of lulc codes, 4)
+        sorted_lucodes (list[int]): List of LULC codes sorted from smallest 
+            to largest. These correspond to the rows of ``ratio_lookup``.
+
+    Returns:
+        2D numpy array with the same shape as ``lulc_array`` and 
+        ``soil_group_array``. Each value is the corresponding ratio for that
+        LULC code x soil group pair.
+    """
+    sorted_soil_groups = [1, 2, 3, 4]
+    # the index of each soil group in the sorted soil groups array
+    soil_group_index = numpy.digitize(soil_group_array, sorted_soil_groups, 
+        right=True)
+    # the index of each lucode in the sorted lucodes array
+    lulc_index = numpy.digitize(lulc_array, sorted_lucodes, right=True)
+    output_ratio_array = ratio_lookup[lulc_index, soil_group_index]
+    return output_ratio_array
+
+
+def volume_op(ratio_array, precip_array, precip_nodata, pixel_area):
+    """Calculate array of volumes (retention or infiltration) from arrays 
+    of precipitation values and stormwater ratios. This is meant to be used
+    with raster_calculator.
+
+    Args:
+        ratio_array (numpy.ndarray): 2D array of stormwater ratios. Assuming
+            that its nodata value is the global NODATA.
+        precip_array (numpy.ndarray): 2D array of precipitation amounts 
+            in millimeters/year
+        precip_nodata (float): nodata value for the precipitation array
+        pixel_area (float): area of each pixel in m^2
+    
+    Returns:
+        2D numpy.ndarray of precipitation volumes in m^3/year
+    """
+    volume_array = numpy.full(ratio_array.shape, NODATA, dtype=float)
+    valid_mask = (
+        (ratio_array != NODATA) & 
+        (precip_array != precip_nodata))
+
+    # precipitation (mm/yr) * pixel area (m^2) * 
+    # 0.001 (m/mm) * ratio = volume (m^3/yr)
+    volume_array[valid_mask] = (
+        precip_array[valid_mask] *
+        ratio_array[valid_mask] *
+        pixel_area * 0.001)
+    return volume_array
+
+
+def avoided_pollutant_load_op(lulc_array, lulc_nodata, retention_volume_array, 
+        sorted_lucodes, emc_array):
+    """Calculate avoided pollutant loads from LULC codes retention volumes.
+    This is intented to be used with pygeoprocessing.raster_calculator.
```

Comment:
Also kind of concerned about validity here... in my test data using 2.5 mg nitrogen/L, 100 mm annual precipitation, over a roughly 40x60km area, I get 492,895 kg avoided nitrogen load. Are there really over 500 tons of nitrogen available to dissolve in rain each year? Idk